libs/libcommon/src/libcommon/viewer_utils/asset.py

```diff
@@ -79,48 +86,74 @@ class AudioSource(TypedDict):
 SupportedSource = Union[ImageSource, AudioSource]
 
 
+def upload_asset_file(
+    url_dir_path: str,
+    filename: str,
+    file_path: Path,
+    overwrite: bool = True,
+    s3_client: Optional[S3Client] = None,
+    s3_bucket: Optional[str] = None,
+    s3_folder_name: Optional[str] = None,
+) -> None:
+    if s3_client is not None and s3_bucket is not None:
+        object_key = f"{s3_folder_name}/{url_dir_path}/{filename}"
+        create_object = overwrite or not s3_client.exists_in_bucket(s3_bucket, object_key)
+        if create_object:
+            s3_client.upload_to_bucket(str(file_path), s3_bucket, object_key)
+            os.remove(file_path)
+
+
 def create_asset_file(
     dataset: str,
     config: str,
     split: str,
     row_idx: int,
     column: str,
     filename: str,
-    assets_base_url: str,
-    assets_directory: StrPath,
+    storage_options: Union[DirectoryStorageOptions, S3StorageOptions],
     fn: Callable[[Path, str, bool], SupportedSource],
     overwrite: bool = True,
-    # TODO: Once assets and cached-assets are migrated to S3, this parameter is no more needed
-    use_s3_storage: bool = False,
-    s3_client: Optional[S3Client] = None,
-    # TODO: Once assets and cached-assets are migrated to S3, the following parameters dont need to be optional
-    s3_bucket: Optional[str] = None,
-    s3_folder_name: Optional[str] = None,
 ) -> SupportedSource:
-    dir_path, url_dir_path = get_asset_dir_name(
-        dataset=dataset,
-        config=config,
-        split=split,
-        row_idx=row_idx,
-        column=column,
-        assets_directory=assets_directory,
-    )
+    # get url dir path
+    assets_base_url = storage_options.assets_base_url
+    assets_directory = storage_options.assets_directory
+    use_s3_storage = isinstance(storage_options, S3StorageOptions)
+    url_dir_path = get_url_dir_path(dataset=dataset, config=config, split=split, row_idx=row_idx, column=column)
+    src = f"{assets_base_url}/{url_dir_path}/{filename}"
 
-    if not use_s3_storage:
-        file_path = dir_path / filename
-        makedirs(dir_path, ASSET_DIR_MODE, exist_ok=True)
-    else:
-        file_path = Path(assets_directory).resolve() / f"{str(uuid4())}-{filename}"
+    # configure file path
+    file_path = (
+        get_unique_path_for_filename(assets_directory, filename)
+        if use_s3_storage
+        else get_and_create_dir_path(
+            dataset=dataset,
+            config=config,
+            split=split,
+            row_idx=row_idx,
+            column=column,
+            assets_directory=assets_directory,
+        )
+        / filename
+    )
 
-    src = f"{assets_base_url}/{url_dir_path}/{filename}"
+    # create file locally
     asset_file = fn(file_path=file_path, src=src, overwrite=overwrite)  # type: ignore
```

Comment:
what is the issue with the type?

Reply:
Mypy does not recognize common parameters for it: `Unexpected keyword argument "file_path" , Unexpected keyword argument "src" and Unexpected keyword argument "overwrite"` errors appeared.
I should have called just `fn(file_path, src, overwrite)` but  as `save_image` and `save_audio` have a different set of parameters, it failed because of a "duplicate" assignation of parameters.
Honesty, I don't know how to properly solve this in mypy.

Reply:
ok, nevermind